Mean teachers (#144)
creatis-myriad/creatis-myriad.github.io: collections/_posts/2023-07-06-Mean_teachers_are_better_role_models.md

    venue:   "NeurIPS 2017"
pdf: https://arxiv.org/pdf/1703.01780.pdf
---

# Highlights

- Method : The mean teacher model **averages model weigths**, instead of label predictions like the Temporal Ensembling method. Allows for more accurate targets and faster feedback between the student and teacher models.
- Performances : demonstrates **enhanced performance on various datasets** (SVHN, CIFAR-10, and ImageNet) when combined with residual networks. Achieves lower error rates and higher accuracy compared to Temporal Ensembling, even when trained with fewer labeled examples.
- Scalability and applicability : Enables working on **larger datasets**, and training with **fewer labeled examples**.

# Introduction
Limitations of deep learning models :
- require large number of parameters and are prone to overfitting. 
- manual high quality labels for training data is expensive and time consuming.

Interest in semi-supervised learning to use the unlabeled data effectively. 

&rarr; focus on **regularization methods** in **semi-supervised learning**

# Method 

## Applying noise and Ensembling method

> ### **Noise regularization**

Reviewer: ```suggestion
### Noise regularization
```
Just a remark for the overall presentation

- Add noise to the input or intermediate representations of a model
- Helps learning abstract invariances
- **Pushes decision boundaries away from labeled data points**  

Limitations of noise regularization :
- classification cost is **undefined for unlabeled examples** &rarr; not helpful in semi-supervised learning.

> &rarr; The $$\Gamma$$ model to overcome this

Reviewer: ```suggestion
&rarr; The $$\Gamma$$ model aims to overcome this
```

I think it might be y personal preference

> ### **The $$\Gamma$$ model[^1]**

Reviewer: ```suggestion
### The $$\Gamma$$ model[^1]
```

- Evaluates each data point with and without noise, and then applies a consistency cost between the two predictions. 

> - Student teacher models where the **teacher generates targets**, which are then **used by itself as a student for learning**

Reviewer: ```suggestion
- It's a student teacher model: the **teacher generates targets**, which are then **used by itself as a student for learning**
```

Limitations of $$\Gamma$$ model :
- the model itself generates targets, which may be incorrect
- the cost of inconsistency outweighs that of misclassification, preventing the learning of new information
- can lead to a **confirmation bias**

&rarr; Confirmation bias can be mitigated by improving the targets. One way is illustrated with the $$\Pi$$ model, selecting a different teacher model than the student model

> **NOTE :** The authors mention another way "to choose the perturbation

> of the representations carefully instead of barely applying additive or multiplicative noise" wich is the subject of another paper by Miyato _et al._ (2017)[^2] that will be refered as VAT+EntMin latter. 

Reviewer: ```suggestion
of the representations carefully instead of barely applying additive or multiplicative noise" which is the subject of another paper by Miyato _et al._ (2017)[^2] that will be refered as VAT+EntMin latter. 
```
just a typo here ;)

> ### **The $$\Pi$$ model[^3]**

Reviewer: ```suggestion
### The $$\Pi$$ model[^3]
```

- **noise** is added to the model **during inference**
- results in a "noisy teacher" that can provide **more accurate targets**. 

&rarr; The $$\Pi$$ model can be further improved by Temporal Ensembling (TE)

> ### **Temporal Ensembling (TE)**

Reviewer: ```suggestion
### Temporal Ensembling (TE)
```

- **exponential moving average (EMA)** prediction for each training example is formed by combining the **current version of the model's predictions** with the **predictions made by earlier versions of the model** that evaluated the same example. 
- improves the quality of the prediction 
- better results when those predictions are used as the teacher predictions

> Limitation of Temporal ensembling :

Reviewer: ```suggestion
Limitation of Temporal Ensembling :
```

- Each **target** (prediction) is **updated only once per training epoch**.
- The learned information is **incorporated** into the training process at a relatively **slow pace**.
- The **larger the dataset**, the **longer it takes** for the updates to span the entire dataset.

&rarr; The authors propose the Mean Teacher model that **averages model weights**, instead of predictions like TE, to improve performances of semi-supervised model.

![](/collections/images/mean_teachers/binary_classification_2.jpg)

> A sketch of a binary classification task with two labeled examples (large black dots) and one unlabeled example, demonstrating how the choice of the unlabeled target (blue circle) affects the fitted function (gray curve).

Reviewer: To make things even more clear, you could add the name of the models that correspond to each figure. For example (d) corresponds to $\Pi$.

- (a) A model with no regularization &rarr; free to fit any function that predicts the labeled training examples well  
- (b) A model trained with noisy labeled data (small dots) &rarr; consistent predictions around labeled data points  
- (c) The teacher model (gray curve) is first fitted to the labeled examples, and then left unchanged during the training of the student model. &rarr; Consistency to noise around unlabeled examples provides additional smoothing  
- (d) Noise on the teacher model reduces the bias of the targets without additional training. &rarr; The expected direction of stochastic gradient descent is towards the mean (large blue circle) of individual noisy targets (small blue circles)  
- (e) Ensemble model &rarr; gives an even better expected target

---
## Mean Teacher 

The Mean Teacher is an average of consecutive student models.

![](/collections/images/mean_teachers/mean_teacher_method.jpg)

- Both the student and the teacher model **evaluate the input applying noise (η, η')**

> - The sotmax output is compared with one-hot label using **classification cost** for the student model and **consistency cost** for the teacher 

Reviewer: ```suggestion
- The softmax output is compared with one-hot label using **classification cost** for the student model and **consistency cost** for the teacher 
```
i think it's what you meant

- The teacher model weights are updated using the **EMA weights of the student model**.

> - Both predictions can be used but the **teacher prediction if more likely to be correct**

Reviewer: ```suggestion
- Both predictions can be used but the **teacher prediction is more likely to be correct**
```

- No classification cost is applied when it is a training step with unlabeled example. 

### Consistency cost 

With $$J$$ the consistency cost, as the expected distance between the prediction of the student model (with weights $$θ$$ and noise $$η$$) and the prediction of the teacher model (with weights $$θ'$$ and noise $$η'$$) :

$$ J(θ) = \underset{x,η',η}{E} [ || f (x, θ', η') - f (x, θ, η) ||^2 ] $$

> - Mean squared error as the consistency cost and ramp up its weight from 0 to its final value during the first 80 epochs.

Reviewer: ```suggestion
Mean squared error is applied to the distance to get the consistency cost. The weight of this consistency loss changes during training: it ramps up from 0 to its final value during the first 80 epochs.
```

I think it might be a bit more clear this way

 &rarr; Initially ignoring it and then gradually introducing more and more consistency. 

> ### Difference between the $\Pi$ model, Temporal Ensembling, and Mean teacher 

Reviewer: ```suggestion
### Difference between the $$\Pi$$ model, Temporal Ensembling, and Mean teacher 
```

- How the teacher predictions are generated :
    - $$\Pi$$ model : $$θ' = θ$$  
    - TE : approximates $$f (x, θ', η')$$ with a **weighted average of successive predictions**
    - Mean Teacher defines $$\underset{t}{θ'}$$ at training step as the **EMA of successive $$\theta$$ weights**, where $$\alpha$$ is a smoothing coefficient hyperparameter : 

$$ \underset{t}{θ'} = α \underset{t-1}{θ'} + (1 - α)\underset{t}{θ} $$

- Weights $$\theta'$$ :
    - $$\Pi$$ model : applies training to it
    - TE and Mean teacher : treat it as a constant with regards to **optimization**

### Type of noise 
- Random translations and horizontal flips of the input images
- Gaussian noise on the input layer
- Dropout applied within the network

### Loss
- Student : cross-entropy loss 

> - Teacher : consitency loss multiplied by an importance weight 

Reviewer: ```suggestion
- Teacher : consistency loss multiplied by an importance weight 
```

## Data
**Street View House Numbers (SVHN)** : 
- 32x32 pixel RGB images belonging to ten different classes
- close-up of a house number, and the class represents the identity of the digit at the center. 

> - 73257 training samples and 26032 test samples.

Reviewer: ```suggestion
- 73,257 training samples and 26,032 test samples.
```

![](/collections/images/mean_teachers/SVHN.jpg)

**CIFAR-10** : 
- 32x32 pixel RGB images belonging to ten different classes
- natural image belonging to a class such as horses, cats, cars and airplanes etc

> - 50000 training samples and 10000 test samples.

Reviewer: ```suggestion
- 50,000 training samples and 10,000 test samples.
```

![](/collections/images/mean_teachers/CIFAR-10.jpg)

# Results

> ## Comparisons of the state-of-the-art methods : 

Reviewer: ```suggestion
## Comparisons with the state-of-the-art methods:
```

![](/collections/images/mean_teachers/table_1.jpg)

**Mean Teacher improves test accuracy** over the $$\Pi$$ model and Temporal Ensembling on semi-supervised SVHN tasks. 

![](/collections/images/mean_teachers/table_2.jpg)

Mean Teacher also improves results on CIFAR-10 over our baseline $$\Pi$$ model.

> **NOTE :** The Virtual Adversarial Training (VAT) by [Miyato et al. 2017](https://doi.org/10.48550/arXiv.1704.03976)[2] performs even better than Mean Teacher on the **1000-label SVHN** and the **4000-label CIFAR-10**. VAT and Mean Teacher are complimentary approaches.  
&rarr; Their combination may yield better accuracy than either of them alone.

> ## Error percentage over 10 runs on SVHN with extra unlabeled training data :

Reviewer: ```suggestion
## Error percentage over 10 runs on SVHN with extra unlabeled training data:
```

![](/collections/images/mean_teachers/table_3.jpg)

> ## Effect of using mean teacher : 

Reviewer: ```suggestion
## Effect of using mean teacher: 
```

![](/collections/images/mean_teachers/figure_3.jpg)

- The **EMA-weighted models** (blue and dark gray curves in the bottom row) give **more accurate predictions** than the bare student models (orange and light gray) after an initial period. 
 
    &rarr; EMA-weighted model as the teacher **improves results** in the semi-supervised settings

> - When using **500 labels** **Mean Teacher** **learns faster**, and continues training after the $$\Pi$$ model stops improving. 

Reviewer: ```suggestion
- When using **500 labels** **Mean Teacher** **learns faster**, and continues learning after the $$\Pi$$ model stops improving. 
```

    &rarr; **Mean Teacher uses unlabeled training data more efficiently than the $$\Pi$$ model**

- In the **all-labeled case**, Mean Teacher and the $$\Pi$$ model **behave identically**. With 500k extra unlabeled examples $$\Pi$$ model keeps improving for longer.

     &rarr; **Mean Teacher learns faster**, and eventually **converges to a better result**, but the sheer amount of data appears to offset $$\Pi$$ model’s worse predictions.

## Mean Teacher with residual networks on CIFAR-10 and ImageNet 

Checking if the methods scales to more natural images on ImageNet with 10% of the labels : 

![](/collections/images/mean_teachers/table_4.jpg)

# Conclusions

> Advantages over Temporal Esembling :

Reviewer: ```suggestion
Advantages over Temporal Ensembling :
```

---
layout: review

> title: Mean teachers are better role models: Weight-averaged consistency targets improve semi-supervised deep learning results

Reviewer 2: The title should be within quotes so that it can properly interpreted by jekyll:
```suggestion
title: "Mean teachers are better role models: Weight-averaged consistency targets improve semi-supervised deep learning results"
```

> tags: machine-learning, neural and evolutionary computing

Reviewer 2: Normally, the tags should simply be separated by a whitespace:
```suggestion
tags: machine-learning neural evolutionary-computing
```

> author: Tarvainen Antti, Valpola Harri

Reviewer 2: The `author` field here is to indicate the author of the review (i.e. you), not the authors of the paper :wink: 
```suggestion
author: "Morgane des Ligneris"
```

cite:

>     authors: "Tarvainen et Valpola"

Reviewer 2: You can give the full name of the authors here, since they're not repeated anymore :wink: 
```suggestion
    authors: "Tarvainen Antti, Valpola Harri"
```

    title:   "Mean teachers are better role models: Weight-averaged consistency targets improve semi-supervised deep learning results"

>     venue:   "Neural Information Processing Systems 30 (NIPS 2017) pre-proceedings"

Reviewer 2: I'd simplify this to just NeurIPS 2017.
```suggestion
    venue:   "NeurIPS 2017"
```

pdf: https://arxiv.org/pdf/1703.01780.pdf
---

# Highlights

> - Method : The mean teachers model **averages model weigths**, instead of label predictions like the Temporal Ensembling method. Allows for more accurate targets and faster feedback between the student and teacher models.

Reviewer 2: Here, since you're talking about the method directly, I would use the singular for teachers:
```suggestion
- Method : The mean teacher model **averages model weigths**, instead of label predictions like the Temporal Ensembling method. Allows for more accurate targets and faster feedback between the student and teacher models.
```

- Performances : demonstrates **enhanced performance on various datasets** (SVHN, CIFAR-10, and ImageNet) when combined with residual networks. Achieves lower error rates and higher accuracy compared to Temporal Ensembling, even when trained with fewer labeled examples.

> - Scalability and applicability : Enables working on **larger dataset**, and training with **fewer labeled examples**.

Reviewer 2: I'd say "dataset" here should be plural:
```suggestion
- Scalability and applicability : Enables working on **larger datasets**, and training with **fewer labeled examples**.
```

# Introduction
Limitations of deep learning models :

> - requires large number of parameters and is prone to overfitting. 

Reviewer 2: Adjust conjugation to be about models (plural):
```suggestion
- require large number of parameters and are prone to overfitting. 
```

- manual high quality labels for training data is expensive and time consuming.

Interest in semi-supervised learning to use the unlabeled data effectively. 

&rarr; focus on **regularization methods** in **semi-supervised learning**

# Method 

> ## Apllying noise and Ensembling method

Reviewer 2: Typo:
```suggestion
## Applying noise and Ensembling method
```

> ### **Noise regularization** : 

Reviewer 2: Remove colon for consistency with other subtitles:
```suggestion
### **Noise regularization**
```

- Add noise to the input or intermediate representations of a model
- Helps learning abstract invariances
- **Pushes decision boundaries away from labeled data points**  

Limitations of noise regularization :
- classification cost is **undefined for unlabeled examples** &rarr; not helpful in semi-supervised learning.

&rarr; The $\Gamma$ model to overcome this

> ### **The $\Gamma$ model [1]**

Reviewer 2: The correct formatting of footnotes in Markdown requires a leading caret before the number (both inline and at the bottom of the page) +
Fix LaTeX mode (double dollar signs are needed in jekyll to enable LaTeX mode):
```suggestion
### **The $$\Gamma$$ model[^1]**
```

- Evaluates each data point with and without noise, and then applies a consistency cost between the two predictions. 
- Student teacher models where the **teacher generates targets**, which are then **used by itself as a student for learning**

> Limitations of $\Gamma$ model :

Reviewer 2: Fix LaTeX mode:
```suggestion
Limitations of $$\Gamma$$ model :
```

- the model itself generates targets, which may be incorrect
- the cost of inconsistency outweighs that of misclassification, preventing the learning of new information
- can lead to a **confirmation bias**

> &rarr; Confirmation biais can be mitigated by improving the targets. One way is illustrated with the $\Pi$ model, selecting a different teacher model than the student model

Reviewer 2: Typo + fix LaTeX mode:
```suggestion
&rarr; Confirmation bias can be mitigated by improving the targets. One way is illustrated with the $$\Pi$$ model, selecting a different teacher model than the student model
```

> > **NOTE :** The authors mention another way : "to choose the perturbation

Reviewer 2: No need for the colon between your text and the quote here:
```suggestion
> **NOTE :** The authors mention another way "to choose the perturbation
```

> of the representations carefully instead of barely applying additive or multiplicative noise" wich is the subject of another paper of [Miyato et al. 2017](https://doi.org/10.48550/arXiv.1704.03976)[2] that will be refered as VAT+EntMin latter. 

Reviewer 2: Fix footnote formatting (the link to the paper should go in the footnote at the bottom):
```suggestion
of the representations carefully instead of barely applying additive or multiplicative noise" wich is the subject of another paper by Miyato _et al._ (2017)[^2] that will be refered as VAT+EntMin latter. 
```

> ### **The $\Pi$ model [3]**

Reviewer 2: Fix LaTeX mode + footnote formatting:
```suggestion
### **The $$\Pi$$ model[^3]**
```

- **noise** is added to the model **during inference**

> - result in a "noisy teacher" that can provide **more accurate targets**. 

Reviewer 2: Conjugation:
```suggestion
- results in a "noisy teacher" that can provide **more accurate targets**. 
```

> &rarr; The $\Pi$ model can be further improved by Temporal Ensembling (TE)

Reviewer 2: Fix LaTeX mode:
```suggestion
&rarr; The $$\Pi$$ model can be further improved by Temporal Ensembling (TE)
```

### **Temporal Ensembling (TE)**
- **exponential moving average (EMA)** prediction for each training example is formed by combining the **current version of the model's predictions** with the **predictions made by earlier versions of the model** that evaluated the same example. 
- improves the quality of the prediction 
- better results when those predictions are used as the teacher predictions

Limitation of Temporal ensembling :
- Each **target** (prediction) is **updated only once per training epoch**.
- The learned information is **incorporated** into the training process at a relatively **slow pace**.
- The **larger the dataset**, the **longer it takes** for the updates to span the entire dataset.

&rarr; The authors propose the Mean Teacher model that **averages model weights**, instead of predictions like TE, to improve performances of semi-supervised model.

![](/collections/images/mean_teachers/binary_classification_2.jpg)

A sketch of a binary classification task with two labeled examples (large black dots) and one unlabeled example, demonstrating how the choice of the unlabeled target (blue circle) affects the fitted function (gray curve).

> (a) A model with no regularization &rarr; free to fit any function that predicts the labeled training examples well  
> (b) A model trained with noisy labeled data (small dots) &rarr; consistent predictions around labeled data points  
> (c) The teacher model (gray curve) is first fitted to the labeled examples, and then left unchanged during the training of the student model. &rarr; Consistency to noise around unlabeled examples provides additional smoothing  
> (d) Noise on the teacher model reduces the bias of the targets without additional training. &rarr; The expected direction of stochastic gradient descent is towards the mean (large blue circle) of individual noisy targets (small blue circles)  
> (e) Ensemble model &rarr; gives an even better expected target

Reviewer 2: Markdown does not support alphabetical lists, so we kind of have to work around that. But I tested things locally, and I think it's much easier to read this section if it is formatted as a normal list:
```suggestion
- (a) A model with no regularization &rarr; free to fit any function that predicts the labeled training examples well  
- (b) A model trained with noisy labeled data (small dots) &rarr; consistent predictions around labeled data points  
- (c) The teacher model (gray curve) is first fitted to the labeled examples, and then left unchanged during the training of the student model. &rarr; Consistency to noise around unlabeled examples provides additional smoothing  
- (d) Noise on the teacher model reduces the bias of the targets without additional training. &rarr; The expected direction of stochastic gradient descent is towards the mean (large blue circle) of individual noisy targets (small blue circles)  
- (e) Ensemble model &rarr; gives an even better expected target
```

---
## Mean Teacher 

The Mean Teacher is an average of consecutive student models.

![](/collections/images/mean_teachers/mean_teacher_method.jpg)

- Both the student and the teacher model **evaluate the input applying noise (η, η')**
- The sotmax output is compared with one-hot label using **classification cost** for the student model and **consistency cost** for the teacher 

> - The teacher model weight are upadted using the **EMA weights of the student model**.

Reviewer 2: Plural + typo:
```suggestion
- The teacher model weights are updated using the **EMA weights of the student model**.
```

> - Both prediction can be used but the **teacher prediction if more likely to be correct**

Reviewer 2: Plural:
```suggestion
- Both predictions can be used but the **teacher prediction if more likely to be correct**
```

> - No classification cost is applied when it is a training step with unlabeled exemple. 

Reviewer 2: Typo:
```suggestion
- No classification cost is applied when it is a training step with unlabeled example. 
```

> ### Concistency cost 

Reviewer 2: Typo:
```suggestion
### Consistency cost 
```

    &rarr; **Mean Teacher uses unlabeled training data more efficiently than the $\Pi$ model**

- In the **all-labeled case**, Mean Teacher and the $\Pi$ model **behave identically**. With 500k extra unlabeled examples $\Pi$ model keeps improving for longer.

     &rarr; **Mean Teacher learns faster**, and eventually **converges to a better result**, but the sheer amount of data appears to offset $\Pi$ model’s worse predictions.

## Mean Teacher with residual networks on CIFAR-10 and ImageNet 

Checking if the methods scales to more natural images on ImageNet with 10% of the labels : 

![](/collections/images/mean_teachers/table_4.jpg)

# Conclusions

Advantages over Temporal Esembling :
- **more accurate target labels**, which enables **faster feedback loop** between the student and the teacher model
- scales to **large datasets** and online learning.

To conclude : 
- The success of consistency regularization depends on the quality of teacher-generated targets.
- Mean Teacher and Virtual Adversarial Training represent
two ways of exploiting this principle.

    &rarr; Their combination may yield even better targets. 

# References

> [1]: Rasmus, Antti, Berglund, Mathias, Honkala, Mikko, Valpola, Harri, and Raiko, Tapani. Semi-supervised Learning with Ladder Networks. In Cortes, C., Lawrence, N. D., Lee, D. D., Sugiyama, M., and Garnett, R. (eds.), Advances in Neural Information Processing Systems 28, pp. 3546–3554. Curran Associates, Inc., 2015  
> 
> [2]: Miyato, Takeru, Maeda, Shin-ichi, Koyama, Masanori, and Ishii, Shin. Virtual Adversarial Training: a Regularization Method for Supervised and Semi-supervised Learning. arXiv:1704.03976 [cs, stat], April 2017. arXiv: 1704.03976.
> 
> [3] Laine, Samuli and Aila, Timo. Temporal Ensembling for Semi-Supervised Learning. arXiv:1610.02242 [cs], October 2016. arXiv: 1610.02242.

Reviewer 2: Fix Markdown formatting of footnotes:
```suggestion
[^1]: Rasmus, Antti, Berglund, Mathias, Honkala, Mikko, Valpola, Harri, and Raiko, Tapani. Semi-supervised Learning with Ladder Networks. In Cortes, C., Lawrence, N. D., Lee, D. D., Sugiyama, M., and Garnett, R. (eds.), Advances in Neural Information Processing Systems 28, pp. 3546–3554. Curran Associates, Inc., 2015  

[^2]: Miyato, Takeru, Maeda, Shin-ichi, Koyama, Masanori, and Ishii, Shin. Virtual Adversarial Training: a Regularization Method for Supervised and Semi-supervised Learning. arXiv:1704.03976 [cs, stat], April 2017. arXiv: 1704.03976.

[^3] Laine, Samuli and Aila, Timo. Temporal Ensembling for Semi-Supervised Learning. arXiv:1610.02242 [cs], October 2016. arXiv: 1610.02242.
```

> [3] Laine, Samuli and Aila, Timo. Temporal Ensembling for Semi-Supervised Learning. arXiv:1610.02242 [cs], October 2016. arXiv: 1610.02242.

Reviewer 2: Another thing I suggest is to shorten your references (e.g. only first author + _et al._, conference/journal + year), but to insert links directly to the papers. That way it's easier for readers to follow up on the references you include, w/o needing to include to much details in the footnotes.

### Type of noise 
- Random translations and horizontal flips of the input images
- Gaussian noise on the input layer
- Dropout applied within the network

### Loss
- Student : cross-entropy loss 
- Teacher : consistency loss multiplied by an importance weight 

## Data
**Street View House Numbers (SVHN)** : 
- 32x32 pixel RGB images belonging to ten different classes
- close-up of a house number, and the class represents the identity of the digit at the center. 
- 73,257 training samples and 26,032 test samples.

![](/collections/images/mean_teachers/SVHN.jpg)

**CIFAR-10** : 
- 32x32 pixel RGB images belonging to ten different classes
- natural image belonging to a class such as horses, cats, cars and airplanes etc
- 50,000 training samples and 10,000 test samples.

![](/collections/images/mean_teachers/CIFAR-10.jpg)

# Results

## Comparisons with the state-of-the-art methods:

![](/collections/images/mean_teachers/table_1.jpg)

**Mean Teacher improves test accuracy** over the $$\Pi$$ model and Temporal Ensembling on semi-supervised SVHN tasks. 

![](/collections/images/mean_teachers/table_2.jpg)

Mean Teacher also improves results on CIFAR-10 over our baseline $$\Pi$$ model.

> > **NOTE :** The Virtual Adversarial Training (VAT) by [Miyato et al. 2017](https://doi.org/10.48550/arXiv.1704.03976)[2] performs even better than Mean Teacher on the **1000-label SVHN** and the **4000-label CIFAR-10**. VAT and Mean Teacher are complimentary approaches.  

Reviewer 2: Fix the already existing footnote rather than inline the reference:
```suggestion
> **NOTE :** The Virtual Adversarial Training (VAT) by Miyato et al. 2017[^2] performs even better than Mean Teacher on the **1000-label SVHN** and the **4000-label CIFAR-10**. VAT and Mean Teacher are complimentary approaches.  
```
